### PR TITLE
Support ValueType override.

### PR DESCRIPTION
**Pull request by jkohen**
Sorry for the long change. This is mostly replacing one type for another.

### CODE REVIEW

File: metadata/cache.go [hunk 1]
```diff
@@ -50,35 +51,56 @@ const DefaultEndpointPath = "api/v1/targets/metadata"
...
 		client:         client,
-		staticMetadata: map[string]scrape.MetricMetadata{},
-		metadata:       map[string]*metadataEntry{},
+		staticMetadata: map[string]*Entry{},
+		metadata:       map[string]*cacheEntry{},
 		seenJobs:       map[string]struct{}{},
 	}
 	for _, m := range staticMetadata {
-		c.staticMetadata[m.Metric] = m
+		c.staticMetadata[m.MetricMetadata.Metric] = m
```
[the remark this answers is not in the document]
Comment (jkohen) [reply metadata/cache.go]: Done. I went one step beyond and flattened the `Entry` type, so it doesn't rely on `MetricMetadata`.

File: retrieval/transform.go [hunk 2]
```diff
@@ -353,3 +354,10 @@ func histogramLabelsEqual(a, b tsdbLabels.Labels) bool {
 	// If one label set still has labels left, they are not equal.
 	return i == len(a) && j == len(b)
 }
+
+func buildTypedValue(valueType metric_pb.MetricDescriptor_ValueType, v float64) *monitoring_pb.TypedValue {
+	if valueType == metric_pb.MetricDescriptor_INT64 {
```
[the remark this answers is not in the document]
Comment (jkohen) [reply retrieval/transform.go]: The fall-back behavior is intended. Added comment.

File: retrieval/series_cache_test.go [hunk 3]
```diff
@@ -546,8 +547,8 @@ func TestSeriesCache_RenameMetric(t *testing.T) {
 		},
 	}
 	metadataMap := metadataMap{
-		"job1/inst1/metric1": &scrape.MetricMetadata{Type: textparse.MetricTypeGauge, Metric: "metric1"},
-		"job1/inst1/metric2": &scrape.MetricMetadata{Type: textparse.MetricTypeGauge, Metric: "metric2"},
+		"job1/inst1/metric1": metadata.NewEntry("metric1", textparse.MetricTypeGauge, metric_pb.MetricDescriptor_DOUBLE, ""),
+		"job1/inst1/metric2": metadata.NewEntry("metric2", textparse.MetricTypeGauge, metric_pb.MetricDescriptor_DOUBLE, ""),
```
[the remark this answers is not in the document]
Comment (jkohen) [reply retrieval/series_cache_test.go]: Done, in transform_test.go. Is there a case I'm missing here?

File: metadata/cache.go [hunk 4]
```diff
@@ -52,23 +51,19 @@ const DefaultEndpointPath = "api/v1/targets/metadata"
...
-func (e *Entry) MetricType() textparse.MetricType {
-	return e.MetricMetadata.Type
+	Metric     string
+	MetricType textparse.MetricType
+	ValueType  metric_pb.MetricDescriptor_ValueType
+	Help       string
 }
 
 // NewEntry returns a new metadata entry.
 func NewEntry(metric string, metricType textparse.MetricType, valueType metric_pb.MetricDescriptor_ValueType, help string) *Entry {
```
Comment (davidbtucker) [inline metadata/cache.go]: Optional: Omit this constructor.

File: cmd/stackdriver-prometheus-sidecar/main_test.go [hunk 5]
```diff
@@ -171,3 +179,92 @@ func TestParseWhitelists(t *testing.T) {
 		})
 	}
 }
+
+func TestProcessFileConfig(t *testing.T) {
+	mustParseMetricSelector := func(input string) []*labels.Matcher {
+		if m, err := promql.ParseMetricSelector(input); err != nil {
+			t.Fatalf("bad test input %v: %v", input, err)
+			return nil
```
Comment (davidbtucker) [inline cmd/stackdriver-prometheus-sidecar/main_test.go]: Nit: Could omit the "return nil" and else.

File: cmd/stackdriver-prometheus-sidecar/main_test.go [hunk 6]
```diff
@@ -171,3 +179,92 @@ func TestParseWhitelists(t *testing.T) {
...
+			if diff := cmp.Diff(tt.renameMappings, renameMappings); diff != "" {
+				t.Errorf("renameMappings mismatch: %v", diff)
+			}
+			if diff := cmp.Diff(tt.staticMetadata, staticMetadata); diff != "" {
+				t.Errorf("staticMetadata mismatch: %v", diff)
+			}
+			if diff := cmp.Diff(tt.aggregations, aggregations); diff != "" {
+				t.Errorf("aggregations mismatch: %v", diff)
+			}
+			if (tt.err != nil && err != nil && tt.err.Error() != err.Error()) ||
```
Comment (davidbtucker) [inline cmd/stackdriver-prometheus-sidecar/main_test.go]: Probably err checking should be before the other diffs.

File: cmd/stackdriver-prometheus-sidecar/main_test.go (hunk 5 above)
Comment (jkohen) [reply cmd/stackdriver-prometheus-sidecar/main_test.go]: Done.

File: metadata/cache.go (hunk 4 above)
Comment (jkohen) [reply metadata/cache.go]: Done. That was a lot of work, hopefully for the better :)

**Pull request merged**